### ISSUE TITLE
HARPOON 2.0

### ISSUE THREAD

**Issue by ThePrimeagen** [ThePrimeagen/harpoon#322]
Harpoon has always been a public source more than open source

i want to make Harpoon 2.0 to make everything you all want.  PLEASE LET ME KNOW THE THINGS YOU NEED TO CHANGE ABOUT Harpoon.

the problem is that Harpoon requires too much integration for any changes, i want to make harpoon generic retriever

i'll start creating harpoon 2.0 in the next couple of weeks

**Comment by nikfp** [comment]
scoping a harpoon list to a git branch could be nice

**Comment by willothy** [comment]
If there are terminal features in the new version (though I think you've mentioned wanting to remove them) it would be nice to have more control over how the terminal commands get dispatched. This way users can make use of their preferred multiplexer, and the burden of maintaining integrations falls less on Harpoon and more on extensions / user configs.

**Comment by ThePrimeagen** [comment]
@willothy 2 things
* i want to completely remove the idea of terminal features
* i want to make harpoon work in such a way that _ANY TERMINAL_ can have commands issued to

i want to create an extensible architecture

**Comment by matu3ba** [comment]
1. make it a Entity Component System like [overseer](https://github.com/stevearc/overseer.nvim) and/or ~~steal~~ reuse the bits. The design of overseer is excellent, but personally I dont like the interface choice with 1. toggle overview, 2. jump into overview, 3. select item, 4. select action, 5. enter
  * the temporary buffer functionality of overseer is excellent except for `:q!`/`ZQ` closing the content
  * terminal functionality could remain and adding any should be avoided unless for extracting autocompletions
2. provide the user an api to access to the underlying (id, buffer, path) and the same for terminal
3. slider from left, right, top or bottom for path actions ie for applying multiple filters is my preference, but overall the popup should be never needed to prevent losing context
4. provide lua api + commands to select local commands and/or how to invoke those commands externally for non-vim users
5. [Preconfigured Terminal Commands](https://github.com/ThePrimeagen/harpoon#preconfigured-terminal-commands), sounds like windows-specific, since the standard on unix seems to be direnv or not?
6. Outline your planned design and/or make a writeup or alike format.
7. once the baseline for the ECS is done, consider adding git worktree as one implementation. Others could be inspired by https://github.com/cbochs/grapple.nvim#project-scopes.

Feature wise, I think less is more unless you plan to do buffer-annotations for serialization and deserialization of complete system states. In that case, you can take a look at my messarounds in https://github.com/matu3ba/libbuf.nvim.

**Comment by ThePrimeagen** [comment]
@nikfp 

what about `key()` as an abstract function.

* defaults to `cwd()`
* allow for override (which could then have `git_branch()` + `cwd()`)

**Comment by ThePrimeagen** [comment]
> 1. make it a Entity Component System like [overseer](https://github.com/stevearc/overseer.nvim) and/or ~steal~ reuse the bits. The design of overseer is excellent, but personally I dont like the interface choice with 1. toggle overview, 2. jump into overview, 3. select item, 4. select action, 5. enter
> 
> * the temporary buffer functionality of overseer is excellent except for `:q!`/`ZQ` closing the content
> * terminal functionality could remain and adding any should be avoided unless for extracting autocompletions
> 
> 2. provide the user an api to access to the underlying (id, buffer, path) and the same for terminal
> 3. slider from left, right, top or bottom for path actions ie for applying multiple filters is my preference, but overall the popup should be never needed to prevent losing context
> 4. provide lua api + commands to select local commands and/or how to invoke those commands externally for non-vim users
> 5. [Preconfigured Terminal Commands](https://github.com/ThePrimeagen/harpoon#preconfigured-terminal-commands), sounds like windows-specific, since the standard on unix seems to be direnv or not?
> 6. Outline your planned design and/or make a writeup or alike format.
> 7. once the baseline for the ECS is done, consider adding git worktree as one implementation. Others could be inspired by https://github.com/cbochs/grapple.nvim#project-scopes.
> 
> Feature wise, I think less is more unless you plan to do buffer-annotations for serialization and deserialization of complete system states. In that case, you can take a look at my messarounds in https://github.com/matu3ba/libbuf.nvim.

i honestly don't understand about 1/3 of this

point 4 seems really confusing... why would i even have an api outside of vim?  this is a vim plugin.  its not a stand alone library.
point 6 will be done on next programming stream

**Comment by catgoose** [comment]
I only use harpoon to cycle between a few files at once.  My bindings:

<leader>a: Add a file
<leader>d: Delete file
<leader>l: Show harpoon
[[ and ]]: prev and next file

I use other neovim plugins and a tmux plugin to dispatch commands.  I don't really need harpoon for that.

**Comment by willothy** [comment]
> * i want to completely remove the idea of terminal features
> * i want to make harpoon work in such a way that _ANY TERMINAL_ can have commands issued to
> 
> i want to create an extensible architecture

That sounds perfect, a system similar to `telescope` or `overseer` in that sense would be very useful.

**Comment by matu3ba** [comment]
> why would i even have an api outside of vim?

Sorry for not being explicit on this one. Afaik, there is no way to log the command execution and the bits to read a command under cursor and logging in lua are also missing (say to read commands from/to command database or local file). 
```lua
lua require("harpoon.term").sendCommand(1, 1)           -- sends command 1 to term 1
```

**Comment by nikfp** [comment]
> @nikfp
> 
> what about `key()` as an abstract function.
> 
>     * defaults to `cwd()`
> 
>     * allow for override (which could then have `git_branch()` + `cwd()`)

sounds like a simple and elegant API to this.

**Comment by otavioschwanck** [comment]
merge this: https://github.com/ThePrimeagen/harpoon/pull/300

**Comment by habanerotostitos** [comment]
not sure if im just dumb since no one mentioned this, but no longer existing files stay in the harpoon list. feels kind of awkward. it sends you to an empty buffer. checking if files exist on start and doing garbage clean up for no longer existing files seems cool. or an option to do so?

aside from that everything works how i want. i dont use any of the terminal/tmux functionality.

**Comment by Danielkonge** [comment]
> not sure if im just dumb since no one mentioned this, but no longer existing files stay in the harpoon list. feels kind of awkward. it sends you to an empty buffer. checking if files exist on start and doing garbage clean up for no longer existing files seems cool. or an option to do so?
> 
> aside from that everything works how i want. i dont use any of the terminal/tmux functionality.

You need to be careful about removing files from the list if you don't keep track of the git branch though. I sometimes have files in one branch that I add to harpoon and want to keep in the list even after using harpoon in another git branch of the same project where said file doesn't exist. 

Also even if you are in a different branch without the file, you might want to use your harpoon shortcut to quickly create and start editing in a corresponding new file in this other branch.

**Comment by habanerotostitos** [comment]
> > not sure if im just dumb since no one mentioned this, but no longer existing files stay in the harpoon list. feels kind of awkward. it sends you to an empty buffer. checking if files exist on start and doing garbage clean up for no longer existing files seems cool. or an option to do so?
> > aside from that everything works how i want. i dont use any of the terminal/tmux functionality.
> 
> You need to be careful about removing files from the list if you don't keep track of the git branch though. I sometimes have files in one branch that I add to harpoon and want to keep in the list even after using harpoon in another git branch of the same project where said file doesn't exist.
> 
> Also even if you are in a different branch without the file, you might want to use your harpoon shortcut to quickly create and start editing in a corresponding new file in this other branch.

yea true. i think if its an option than you can get the combination of settings thats ideal for everyone.

or maybe even better, if theres just as an easy :harpoon garbagecollect we can call in our own autocmds at our own frequency if we want it only at start, only whenever the menu is opened, or more frequent than that, it is easy to control for each user.

**Comment by nikfp** [comment]
@habanerotostitos this is where having a list keyed to a cwd and git branch would be useful. If you want to clean the state of a list while on a particular branch it will be as easy as opening the list and deleting what you don't need. However you could switch from a feature branch to a bugfix, do what you need with a discrete list, and then switch back and your previous list would be intact.

**Comment by Yvey** [comment]
I am still new to this Vim/Neovim stuff.  
Adding file with an optional line position to harpoon.
i.e.
1 - file2.c  , line 28
2 - main.c  , line 11
3 - file2.c  , line 257
4 - header.c  , no line saved

I am not sure how useful it is to others.
I often find the need to refactor a section of a code and end up changing the whole section of a function when doing my uni project, I probably should change my workflow and start looking at what other c/cpp devs do.
or maybe there is already a function like that in nvim that I am not aware of... :|

**Comment by adampresley** [comment]
I don't really use the terminal stuff, so I primarily use harpoon for keeping a *short* list of open files in my current working *context* (think the "stuff or section I'm working in now"). So I agree with @victorscomma in the comment above, that it would be interesting to be able to have a "list" of contexts. 

For example, I've got 4 files marked in harpoon related to "project details" code files.

* ProjectHandlers.go
* ProjectService.go
* project-details.tmpl
* project-details.js

Then let's say I need to switch context to "user management", with a different set of files/harpoon marks. Perhaps if there was a way to pin these sets of files to different "lists" and easily switch between them.

That *could* be useful.

**Comment by otavioschwanck** [comment]
> I am still new to this Vim/Neovim stuff. Adding file with an optional line position to harpoon. i.e. 1 - file2.c , line 28 2 - main.c , line 11 3 - file2.c , line 257 4 - header.h , no line saved
> 
> I am not sure how useful it is to others. I often find the need to refactor a section of a code and end up changing the whole section of a function when doing my uni project, I probably should change my workflow and start looking at what other c/cpp devs do. or maybe there is already a function like that in nvim that I am not aware of... :|

instead of the line number, could pick the context with tree-sitter, like:

file.c:(f)find_user
file.c:(v)name

**Comment by mikejmcguirk** [comment]
I have a custom function written where, if I am in a buffer that is marked, I can remove the mark then rebuild the marks list so there isn't an empty file, and then refresh the tab line. Something like that would be a good functionality to just have as a default I could put a keymap on.

https://github.com/ThePrimeagen/harpoon/pull/309 - This Pull request addresses an issue with file normalization in Windows that prevents the tabline from displaying properly

**Comment by arnevm123** [comment]
It would be nice to have a way to truncate the path.
When working in a large monorepo all marks sometimes just have the same start and that can be a bit annoying.

**Comment by juayhee** [comment]
Thanks very much for harpoon. It is amazing. 

A few features I thought I would have liked when using Harpoon.

1. My use case is that I always navigate back to a file I want using the pop-up menu instead of the 1/2/3 key directly. Because I can't remember which I put what on. 

I would like more qol for the popup
- hovering over an item in the popup opens a preview window like telescope, and you can move around in it with ctrl d/u (like telescope)
- Being able to annotate or rename the list items instead of it being just the file name

2. I commonly have a set of files that I am going between (for example, header files and their source files). Frequently I find I might want to start a new header/source pair of files and work on them. But I don't want to clear the harpoon list each time. I would like to maintain separate 'working sets' of harpoon files

**Comment by otavioschwanck** [comment]
Harpoon code have those actions to anyfile:

- Rename
- Delete
- Move (or just use rename to move)

Those actions could work on anyfile, but if the target file is on harpoon, also change on harpoon.

**Comment by ch4uz** [comment]
If there are no marked files, harpoon will automatically create marks (max 5 marks) based on [frecency](https://en.wikipedia.org/wiki/Frecency)

**Comment by ThePrimeagen** [comment]
These have been very incredible suggestions.

Be prepared for November 3rd, that is when the rewrite will happen, on stream, for 12 hours or until done

**Comment by PhilShaughnes** [comment]
kinda late, but adding my 2 cents:
* keep it simple - a way to add, reference, and cycle important files I'm working on NOW
* no terminal, dispatch, etc etc
* making it easy to populate, reset, modify, maybe even save that list is great
* the "it's just a buffer" is great. 
* integrating quickfix (easy to populate with lua-fzf or telescope) and something like :cdo would be nifty and maybe also swell.

**Comment by sjclayton** [comment]
Another thing, harpoon should have an option to enable/disable remembering the last edit in a file when opening, because if a user has an autocmd or something else setup to do this for all files, harpoon doing it as well messes things up when transitioning between harpoon marked files and non-harpoon marked files.

**Comment by mvargasmoran** [comment]
Anime Girl Icons?
ok, PHP, JS, React icons for the files in the list.

or just a waifu background in the floating window.
well, whatever waifu you want, from MrMomMusic would suffice.

**Comment by Flygrounder** [comment]
I would love to be able to first select a project that has marks and then open a marked file in that project. Would make my life so much easier.

**Comment by mlp1802** [comment]
Already made a feature request for this..multiple named harpoon lists..instead of just one list, this would take it to the next level.

**Comment by TangerineGameDev** [comment]
I ask for one feature!

Allow for a max amount of added files/buffers.
Whenever the buffer overflows, then use a queue approach (first-in, first-out).

For example:

- I have 5 set as a max amount of files.
- Harpoon shows: File 1 (Oldest), File 2, File 3, File 4, File 5 (Newest)
- Add a file here (File 6) causing the max number of buffers to overflow
- Now Harpoon looks like this: File 2 (Oldest), File 3, File 4, File 5, File 6 (Newest)

We could even try something else, but a way to drop and replace files with newer files would be appreciated!

**Comment by mrs4ndman** [comment]
May I suggest per-tab Harpoon lists?

**Comment by TangerineGameDev** [comment]
> line

Already mentioned @michaelPotter

**Comment by TangerineGameDev** [comment]
> I am still new to this Vim/Neovim stuff. Adding file with an optional line position to harpoon. i.e. 1 - file2.c , line 28 2 - main.c , line 11 3 - file2.c , line 257 4 - header.h , no line saved
> 
> I am not sure how useful it is to others. I often find the need to refactor a section of a code and end up changing the whole section of a function when doing my uni project, I probably should change my workflow and start looking at what other c/cpp devs do. or maybe there is already a function like that in nvim that I am not aware of... :|

Already mentioned @michaelPotter

**Comment by otavioschwanck** [comment]
> Anime Girl Icons? ok, PHP, JS, React icons for the files in the list.
> 
> or just a waifu background in the floating window. well, whatever waifu you want, from MrMomMusic would suffice.
> 
> <img alt="image" width="695" src="https://user-images.githubusercontent.com/5852003/276749921-050fc474-d180-4f23-b38c-98f78b9775fb.png">

https://github.com/ThePrimeagen/harpoon/pull/300 Already did here, just need to merge

**Comment by ThePrimeagen** [comment]
ok... @victorscomma 

> Some things I think would be cool

Allow us to make more harpoon lists, which can have custom contexts. Allowing maybe something like a global harpoon list. Rather than just the current dir, the global harpoon list can work wherever you are for example.
Allow some means of a dynamic path so we can have things in the harpoon list that don't exist yet.
Options for what happens when you switch to it, i.e currently just switches the active buffer but could also open selected harpoon path on a new split, open as a pop up, etc.
For example if I wanted to harpoon some file related to my health and diet. I want that accessible and jump-to-able from wherever I am currently working. But at the same time I might not want to context switch entirely from my work so being able to set it just a pop up to jot something down quick might make sense.

----

I think we can do most of this.  the contexts may be a bit interesting and i'll have to think about that as that is not very obvious how this would work.  but please feel free to join during Nov 3rd while i rebuild harpoon 2 live on stream.

**Comment by ThePrimeagen** [comment]
> Would be great if you could mark two separate spots in the same file... e.g. for working on really large files or two related functions in the same file

this is ackshually really difficult.  i would recommend local marks for this (ma to mark a and 'a to recall a)

The reason is this, when i close down vim, and your cursor has moved, which of the two entries do i update the cursor to?

to me this is what local marks are great at

**Comment by ThePrimeagen** [comment]
@mrs4ndman 
> May I suggest per-tab Harpoon lists?

i think my idea for how to generate the "key" could work well, but you may find yourself in an annoying situation where the file you want is in tab 2 and you have open up tab 2 to get there or furthermore, you may mind yourself annoyed by trying to remember 8 separate files
 :)

**Comment by ThePrimeagen** [comment]
@TangerineGameDev i think this would be possible with my new idea on lists

> I ask for one feature!

> Allow for a max amount of added files/buffers.
Whenever the buffer overflows, then use a queue approach (first-in, first-out).

> For example:

> I have 5 set as a max amount of files.
Harpoon shows:
File 1 (Oldest), File 2, File 3, File 4, File 5 (Newest)
Add a file here (File 6) causing the max number of buffers to overflow
Now Harpoon looks like this:
File 2 (Oldest), File 3, File 4, File 5, File 6 (Newest)
We could even try something else, but a way to drop and replace files with newer files would be appreciated!

I want to make it so that you can have your own "lists" and have them defined however you would like.  this would potentially involve a callback with `op` and `list` and `args` and you get to chose how you want to store / add / delete

**Comment by ThePrimeagen** [comment]
@otavioschwanck 

> Harpoon code have those actions to anyfile:

> Rename
Delete
Move (or just use rename to move)
Those actions could work on anyfile, but if the target file is on harpoon, also change on harpoon.

these are already available both programmatically and through the UI (just delete or move or manually add a line to the UI and it saves the current state as the list (removes empties))

**Comment by ThePrimeagen** [comment]
@sjclayton 

> Another thing, harpoon should have an option to enable/disable remembering the last edit / cursor location in a file when opening, because if a user has an autocmd or something else setup to do this for all files, harpoon doing it as well messes things up when transitioning between harpoon marked files and non-harpoon marked files.

great idea

**Comment by michaelPotter** [comment]
@ThePrimeagen 
> The reason is this, when i close down vim, and your cursor has moved, which of the two entries do i update the cursor to?

Update whichever one I jumped to last... You'd have to carry some extra data around about jump history but doesn't seem incredibly hard (famous last words)

All good if that's not feasible. I'll give local marks some love

**Comment by mrs4ndman** [comment]
@ThePrimeagen 
> 
> > May I suggest per-tab Harpoon lists?
> 
> i think my idea for how to generate the "key" could work well, but you may find yourself in an annoying situation where the file you want is in tab 2 and you have open up tab 2 to get there or furthermore, you may mind yourself annoyed by trying to remember 8 separate files :)

Yeah, kinda. I was thinking more like fully-independent harpoon lists for each tab. Either way, gonna be watching Nov 3rd how it draws out :heart:

**Comment by spacey-sooty** [comment]
I would like to see Harpoon similar to telescope in the way one can make plugins to extend it. Eg. support neovim marks out of the box then make a plugin api so if someone wants to support something they can

**Comment by paulber33** [comment]
Ohayou gozaimasu, senpai! (´∩｡• ᵕ •｡∩`) I totally get your Harpoon 2.0 vision desu~ (＾▽＾)

So, like, Harpoon has always been, you know, kinda like a public source-chan more than, uh, open source-senpai, but now you wanna give it a kawaii makeover! (≧∇≦) That's super sugoi! (*≧ω≦)

But, uhm, like, the issue is that Harpoon is being a bit of a clingy-senpai right now, needing too much integration, and you wanna make it more like a generic retriever, nya~ (≧◡≦) That's quite the transformation!

Your promise to create Harpoon 2.0 in the next couple of weeks is like a blossoming sakura tree in spring, senpai! 🌸✨

So, lemme get this straight, you wanna change Harpoon, like, drastically into a kawaii generic retriever, but you need some tips and tricks from the community to make it purrfect! (¬‿¬) Nyan-tastic! (^‿^)

I guess it's time to gather our senpais and senseis and brainstorm the way to make Harpoon 2.0 the ultimate, uhm, waifu of retrievers, senpai! (¬‿¬) Let's make it so sugoi that everyone's hearts will go doki-doki~ 💖💖💖

Ganbatte, senpai! (*≧ω≦) We're here to support you in your kawaii quest for Harpoon 2.0! (人 •͈ᴗ•͈) ✨

**Comment by MohamedBassem** [comment]
Maybe adding support for "commenting" out some buffers from the harpoon buffer list by prefixing them with `//` for example?

**Comment by janeusz2000** [comment]
I would love to have a functionality that will allow me to execute unix command on file selected in harpoon terminal. Sometime I need to really jump around the repo in order to execute something and this would save some hustle ;)

**Comment by luigiMinardi** [comment]
Will harpoon 2 use a beautiful UI like the `mason.nvim` or `lazy.nvim`? Not a necessity but it would be nice if we had some beauty in it.

**Comment by devstefancho** [comment]
Feature Request: Highlight Current Buffer Name if Listed in Harpoon

I'm suggesting an enhancement for the harpoon.nvim plugin. It would be highly beneficial to have the buffer's filename highlighted when it's included in the Harpoon list. This would offer immediate visual confirmation and prevent confusion, especially when dealing with files with similar names.

For instance, if the open files are bar.ts, Bar.ts, and baz.ts, and the Harpoon list contains bar.ts, barr.ts, and barz.ts, the highlight would make it instantly clear when the current buffer (bar.ts in this case) is part of the Harpoon list.

This feature would be a significant usability boost for quickly navigating and recognizing files within the Harpoon interface.

**Comment by devstefancho** [comment]
highlighting current buffer file name in harpoon list

this would offer immediate visual confirmation and prevent confusion, especially when dealing with files with similar names.

for example, if the open files are bar.ts, Bar.ts, and baz.ts, and the Harpoon list contains bar.ts, barr.ts, and barz.ts, the highlight would make it instantly clear when the current buffer (bar.ts in this case) is part of the Harpoon list.

**Comment by chrisbirster** [comment]
This is probably more of a skill issue. If I delete or renamed a file that I already have in my harpoon list. The harpoon list will still have the reference to the old file.

**Comment by ThePrimeagen** [comment]
> Ohayou gozaimasu, senpai! (´∩｡• ᵕ •｡∩`) I totally get your Harpoon 2.0 vision desu~ (＾▽＾)
> 
> So, like, Harpoon has always been, you know, kinda like a public source-chan more than, uh, open source-senpai, but now you wanna give it a kawaii makeover! (≧∇≦) That's super sugoi! (*≧ω≦)
> 
> But, uhm, like, the issue is that Harpoon is being a bit of a clingy-senpai right now, needing too much integration, and you wanna make it more like a generic retriever, nya~ (≧◡≦) That's quite the transformation!
> 
> Your promise to create Harpoon 2.0 in the next couple of weeks is like a blossoming sakura tree in spring, senpai! 🌸✨
> 
> So, lemme get this straight, you wanna change Harpoon, like, drastically into a kawaii generic retriever, but you need some tips and tricks from the community to make it purrfect! (¬‿¬) Nyan-tastic! (^‿^)
> 
> I guess it's time to gather our senpais and senseis and brainstorm the way to make Harpoon 2.0 the ultimate, uhm, waifu of retrievers, senpai! (¬‿¬) Let's make it so sugoi that everyone's hearts will go doki-doki~ 💖💖💖
> 
> Ganbatte, senpai! (*≧ω≦) We're here to support you in your kawaii quest for Harpoon 2.0! (人 •͈ᴗ•͈) ✨

What in the actual fuck?

**Comment by giuseppe-g-gelardi** [comment]
How about a harpoon vscode plugin? 

I nearly got fired for installing nvim on my work computer and am stuck using it for work. Would be cool if, you know, I had harpoon there too.

**Comment by ThePrimeagen** [comment]
> How about a harpoon vscode plugin?
> 
> I nearly got fired for installing nvim on my work computer and am stuck using it for work. Would be cool if, you know, I had harpoon there too.

1. already created
2. fired for neovim?  cray

**Comment by ThePrimeagen** [comment]
HARPOON 2 Base is almost done

will be opening it up for community feedback / contributions and we are going from beta/public source to open source and i will ackshually maintain and allow for contributions.  they will be narrow, but it will great

**Comment by rusnasonov** [comment]
Allow working with different vcs

**Comment by minhtien3297** [comment]
> scoping a harpoon list to a git branch could be nice

I use this to save harpoon list to git branch

```   
harpoon:setup({
      settings = {
        save_on_toggle = true,

        -- save harpoon list to git branch
        key = function()
          local handle = io.popen("git branch --show-current 2> /dev/null")
          local git_branch = handle:read("*a") or ""
          handle:close()
          git_branch = string.gsub(git_branch, "\n", "")
          local cwd = vim.fn.getcwd()

          return git_branch .. cwd
        end,
      },
    })
```
